### ISSUE TITLE
Problems with using 2 gadgets in a prio3 type

### ISSUE THREAD

**Issue by MxmUrw** [divviup/libprio-rs#254]
We tried to implement a prio3 type which uses 2 kinds of gadgets, and stumbled upon an `index_out_of_bounds` error.

This might be because we chose the wrong constants for one of `proof_len()`, `joint_rand_len()`, etc.. If so, we would be happy about some guidance with regards to this.

### Details
We created a minimal example type producing our error by modifying the `Sum` type to have a second gadget. It even occurs when the second gadget is not called at all.

Interestingly, the existence of a second gadget alone is not enough to trigger the error: In the test that triggers it, `self.bits` is 16, and the error occurs when the first gadget is called `2 * self.bits` times. It does not occur when there are e.g. `16 + 14` calls.

Note: Of course, when testing the different numbers of calls, we updated all constants in what seemed to be the correct way.

**EDIT**: The number of calls seems to [decide](https://github.com/divviup/libprio-rs/blob/69771d804019ffef8535bf0f005d3fa160a94d7c/src/flp/gadgets.rs#L200) whether the `call_poly_fft` or `call_poly_direct` implementation is called. The error occurs only when the `fft` version is called.

### How to reproduce
The modified `Sum` type that triggers this bug can be found in this [fork](https://github.com/MxmUrw/libprio-rs).
1. Clone that repo.
2. Run the relevant test `cargo test -- test_prio3_sum`.

The output is:
```
> cargo test -- test_prio3_sum
    Finished test [unoptimized + debuginfo] target(s) in 0.06s
     Running unittests (target/debug/deps/prio-53972546cb9089a3)

running 1 test
test vdaf::prio3::tests::test_prio3_sum ... FAILED

failures:

---- vdaf::prio3::tests::test_prio3_sum stdout ----
thread 'vdaf::prio3::tests::test_prio3_sum' panicked at 'index out of bounds: the len is 128 but the index is 128', src/flp/gadgets.rs:171:43
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    vdaf::prio3::tests::test_prio3_sum

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 77 filtered out; finished in 0.00s

error: test failed, to rerun pass '-p prio --lib'
```

**Comment by cjpatton** [comment]
Hi @MxmUrw, thank your for the report. I'm taking a look now. For context, we noticed that using FFT to compute `call_poly()` is more performant than the direct computation for large input sizes.

**Comment by cjpatton** [comment]
I've confirmed the issue, I've created a PR that adds a test case that reproduces it. I'm pretty sure the length of the internal buffer used by `call_poly_fft()` is being computed incorrectly. However I don't understand why we go out-of-bounds when there are two gadgets, but not one. Once I figured it out I will update the PR with the fix.

**Comment by cjpatton** [comment]
The root cause has been identified and the PR has been updated with a fix. @MxmUrw, could you rebase your branch and make sure the issue is resolved on your end?

**Comment by MxmUrw** [comment]
Great, it works now!

**Comment by tgeoghegan** [comment]
Closing as this appears be resolved by #255. Thank you for the thorough report @MxmUrw!

**Comment by tgeoghegan** [comment]
Whoops, I didn't notice that #255 is not merged. Re-opening the issue.
